File: golang/快速幂.go
//快速幂.go

package main

import "fmt"

const MOD = int(1e9 + 7)

func pow(x, n int) int {
	//base 底数
	b := x
	//power 指数
	e := n
	//result 结果
	r := 1
	for e > 0 {
		if e&1 > 0 {
			r *= b
			r %= MOD
		}
		e >>= 1
		b *= b
		b %= MOD
	}
	return r
}

func main() {
	fmt.Println(pow(3, 4))
	//81
	fmt.Println(pow(107, 8))
	//678046174
}

File: python/快速幂.py
# 快速幂.py

MOD = int(1e9 + 7)


def f(x, n, MOD):
    # base 底数
    b = x
    # power 指数
    e = n
    # result 结果
    r = 1
    while e:
        if e & 1:
            r *= b
            r %= MOD
        e >>= 1
        b *= b
        b %= MOD
    return r


x, n = 3, 4
print(pow(x, n, MOD))
print(pow(x, n, MOD) == f(x, n, MOD))
x, n = 107, 8
print(pow(x, n, MOD) == f(x, n, MOD))
